File: src/test/java/com/truveta/opentoken/io/PersonAttributesCSVReaderTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io;

import static org.junit.jupiter.api.Assertions.*;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Map;
import java.util.NoSuchElementException;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class PersonAttributesCSVReaderTest {

    private File tempFile;

    @BeforeEach
    void setUp() throws IOException {
        tempFile = File.createTempFile("test_data", ".csv");
    }

    @AfterEach
    void tearDown() {
        if (tempFile.exists()) {
            tempFile.delete();
        }
    }

    @Test
    void testReadValidCSV() throws Exception {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
            writer.write("RecordId,SocialSecurityNumber,Name\n");
            writer.write("1,123-45-6789,John Doe\n");
            writer.write("2,987-65-4321,Jane Smith\n");
        }

        try (PersonAttributesCSVReader reader = new PersonAttributesCSVReader(tempFile.getAbsolutePath())) {
            assertTrue(reader.hasNext());

            Map<String, String> firstRecord = reader.next();
            assertEquals("1", firstRecord.get("RecordId"));
            assertEquals("123-45-6789", firstRecord.get("SocialSecurityNumber"));
            assertEquals("John Doe", firstRecord.get("Name"));

            assertTrue(reader.hasNext());

            Map<String, String> secondRecord = reader.next();
            assertEquals("2", secondRecord.get("RecordId"));
            assertEquals("987-65-4321", secondRecord.get("SocialSecurityNumber"));
            assertEquals("Jane Smith", secondRecord.get("Name"));

            assertFalse(reader.hasNext());
        }
    }

    @Test
    void testReadEmptyCSV() throws Exception {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
            writer.write("RecordId,SocialSecurityNumber,Name\n");
        }

        try (PersonAttributesCSVReader reader = new PersonAttributesCSVReader(tempFile.getAbsolutePath())) {
            assertFalse(reader.hasNext());
        }
    }

    @Test
    void testHasNext() throws Exception {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
            writer.write("RecordId,SocialSecurityNumber,Name\n");
            writer.write("1,123-45-6789,John Doe\n");
        }

        try (PersonAttributesCSVReader reader = new PersonAttributesCSVReader(tempFile.getAbsolutePath())) {
            assertTrue(reader.hasNext());
            reader.next();
            assertFalse(reader.hasNext());
        }
    }

    @Test
    void testNext() throws Exception {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
            writer.write("RecordId,SocialSecurityNumber,Name\n");
            writer.write("1,123-45-6789,John Doe\n");
        }

        try (PersonAttributesCSVReader reader = new PersonAttributesCSVReader(tempFile.getAbsolutePath())) {
            Map<String, String> record = reader.next();
            assertNotNull(record);
            assertEquals("1", record.get("RecordId"));
            assertEquals("123-45-6789", record.get("SocialSecurityNumber"));
            assertEquals("John Doe", record.get("Name"));
        }
    }

    @Test
    void testClose() throws Exception {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
            writer.write("RecordId,SocialSecurityNumber,Name\n");
            writer.write("1,123-45-6789,John Doe\n");
        }

        PersonAttributesCSVReader reader = new PersonAttributesCSVReader(tempFile.getAbsolutePath());
        reader.close();
        assertThrows(NoSuchElementException.class, () -> reader.next());
    }

    @Test
    void testConstructorThrowsIOException() {
        String invalidFilePath = "non_existent_file.csv";
        assertThrows(IOException.class, () -> new PersonAttributesCSVReader(invalidFilePath));
    }
}

File: lib/python/openlinktoken-cli/src/test/openlinktoken_cli/io/csv/person_attributes_csv_reader_test.py
# SPDX-License-Identifier: MIT

import os
import tempfile

import pytest

from openlinktoken_cli.io.csv.person_attributes_csv_reader import PersonAttributesCSVReader


class TestPersonAttributesCSVReader:
    """Test cases for PersonAttributesCSVReader."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.temp_file = tempfile.NamedTemporaryFile(mode="w", suffix=".csv", delete=False)
        self.temp_file_path = self.temp_file.name
        self.temp_file.close()

    def teardown_method(self):
        """Clean up after each test method."""
        if os.path.exists(self.temp_file_path):
            os.unlink(self.temp_file_path)

    def test_read_valid_csv(self):
        """Test reading a valid CSV file."""
        with open(self.temp_file_path, "w", encoding="utf-8") as f:
            f.write("RecordId,SocialSecurityNumber,FirstName,LastName\n")
            f.write("1,123-45-6789,John,Doe\n")
            f.write("2,987-65-4321,Jane,Smith\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            # Test first record
            first_record = next(reader)
            assert first_record["RecordId"] == "1"
            assert first_record["SocialSecurityNumber"] == "123-45-6789"
            assert first_record["FirstName"] == "John"
            assert first_record["LastName"] == "Doe"

            # Test second record
            second_record = next(reader)
            assert second_record["RecordId"] == "2"
            assert second_record["SocialSecurityNumber"] == "987-65-4321"
            assert second_record["FirstName"] == "Jane"
            assert second_record["LastName"] == "Smith"

            # Test no more records
            with pytest.raises(StopIteration):
                next(reader)

    def test_read_empty_csv(self):
        """Test reading an empty CSV file."""
        with open(self.temp_file_path, "w", encoding="utf-8") as f:
            f.write("RecordId,SocialSecurityNumber,Name\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            with pytest.raises(StopIteration):
                next(reader)

    def test_date_of_birth_alias_maps_to_birth_date(self):
        """Test that DateOfBirth is normalized to the BirthDate field ID."""
        with open(self.temp_file_path, "w", encoding="utf-8") as f:
            f.write("DateOfBirth\n")
            f.write("1980-01-15\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            assert next(reader)["BirthDate"] == "1980-01-15"

    def test_ssn_alias_maps_to_social_security_number(self):
        """Test that SSN is normalized to the SocialSecurityNumber field ID."""
        with open(self.temp_file_path, "w", encoding="utf-8") as f:
            f.write("SSN\n")
            f.write("123-45-6789\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            assert next(reader)["SocialSecurityNumber"] == "123-45-6789"

    def test_iterator_protocol(self):
        """Test iterator protocol."""
        with open(self.temp_file_path, "w", encoding="utf-8") as f:
            f.write("RecordId,SocialSecurityNumber,FirstName,LastName\n")
            f.write("1,123-45-6789,John,Doe\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            # Test that we can iterate
            for record in reader:
                assert record["RecordId"] == "1"
                assert record["SocialSecurityNumber"] == "123-45-6789"
                assert record["FirstName"] == "John"
                assert record["LastName"] == "Doe"
                break

    def test_next(self):
        """Test next method."""
        with open(self.temp_file_path, "w", encoding="utf-8") as f:
            f.write("RecordId,SocialSecurityNumber,FirstName,LastName\n")
            f.write("1,123-45-6789,John,Doe\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            record = next(reader)
            assert record is not None
            assert record["RecordId"] == "1"
            assert record["SocialSecurityNumber"] == "123-45-6789"
            assert record["FirstName"] == "John"
            assert record["LastName"] == "Doe"

    def test_close(self):
        """Test close method."""
        with open(self.temp_file_path, "w", encoding="utf-8") as f:
            f.write("RecordId,SocialSecurityNumber,Name\n")
            f.write("1,123-45-6789,John Doe\n")

        reader = PersonAttributesCSVReader(self.temp_file_path)
        reader.close()

        # After closing, next should raise StopIteration
        with pytest.raises(ValueError):
            next(reader)

    def test_constructor_throws_io_exception(self):
        """Test constructor throws IOError for non-existent file."""
        invalid_file_path = "non_existent_file.csv"
        with pytest.raises(IOError):
            PersonAttributesCSVReader(invalid_file_path)

    def test_row_count_preserves_iteration(self):
        """Counting rows should not prevent subsequent iteration."""
        with open(self.temp_file_path, "w", encoding="utf-8") as f:
            f.write("RecordId,SocialSecurityNumber,FirstName,LastName\n")
            f.write("1,123-45-6789,John,Doe\n")
            f.write("2,987-65-4321,Jane,Smith\n")

        with PersonAttributesCSVReader(self.temp_file_path) as reader:
            assert reader.row_count() == 2

            first_record = next(reader)
            second_record = next(reader)

            assert first_record["RecordId"] == "1"
            assert second_record["RecordId"] == "2"
